pkg/transport/custom_transform_collection.go

```diff
@@ -0,0 +1,216 @@
+
+	// bindingNameToGroupResources tracks the set of GroupResource that each Binding
+	// references. This is so that when the set for a given Binding changes,
+	// for the GroupResources that are no longer in the set, the Binding's Name can
+	// be removed from groupResourceTransformData.bindingsThatCare.
+	bindingNameToGroupResources map[string]sets.Set[metav1.GroupResource]
+}
+
+// groupResourceTransformData is the ingested custom transforms for a given GroupResource
+type groupResourceTransformData struct {
+	bindingsThatCare sets.Set[string /*Binding name*/] // not empty
+	removes          []jsonpath.Query
+}
+
+func newCustomTransformCollection(client controlclient.CustomTransformInterface, getTransformObjects func(indexName, indexedValue string) ([]any, error), enqueue func(any)) *customTransformCollection {
+	return &customTransformCollection{
+		client:                      client,
+		getTransformObjects:         getTransformObjects,
+		enqueue:                     enqueue,
+		grToTransformData:           make(map[metav1.GroupResource]*groupResourceTransformData),
+		ctNameToSpec:                make(map[string]v1alpha1.CustomTransformSpec),
+		bindingNameToGroupResources: make(map[string]sets.Set[metav1.GroupResource]),
+	}
+}
+
+// Call this to process a new version, or deletion, of a CustomTransform.
+func (ctc *customTransformCollection) noteCustomTransform(ctx context.Context, name string, ct *v1alpha1.CustomTransform) {
+	ctc.mutex.Lock()
+	defer ctc.mutex.Unlock()
+	oldSpec, hadSpec := ctc.ctNameToSpec[name]
+	if ct == nil && !hadSpec {
+		return // was absent and is absent; no change
+	}
+	var oldGroupResource, newGroupResource metav1.GroupResource
+	if hadSpec {
+		oldGroupResource = ctSpecGroupResource(oldSpec)
+	}
+	if ct != nil {
+		newGroupResource = ctSpecGroupResource(ct.Spec)
+	}
+	if ct != nil && hadSpec &&
+		oldGroupResource == newGroupResource &&
+		sets.New(oldSpec.Remove...).Equal(sets.New(ct.Spec.Remove...)) {
+		return // unchanged
+	}
+	// GroupResource or Spec.Remove changed
+	logger := klog.FromContext(ctx)
+	if hadSpec {
+		// Invalidate the cached analysis for the relevant GroupResource, if there is one
+		if oldGRTD, hadGRTD := ctc.grToTransformData[oldGroupResource]; hadGRTD {
+			delete(ctc.grToTransformData, oldGroupResource)
+			for bindingName := range oldGRTD.bindingsThatCare {
+				logger.V(5).Info("Enqueuing reference to Binding because of change to CustomTranform", "bindingName", bindingName, "customTransformName", name)
+				ctc.enqueue(bindingName)
+			}
+		}
+	}
+	if ct != nil {
+		ctc.ctNameToSpec[name] = ct.Spec
+	} else {
+		delete(ctc.ctNameToSpec, name)
+	}
+}
+
+// getCustomTransformData returns the groupResourceTransformData to use
+// for the given GroupResource on behalf of the named Binding.
+func (ctc *customTransformCollection) getCustomTransformData(ctx context.Context, groupResource metav1.GroupResource, bindingName string) *groupResourceTransformData {
```

Comment:
Understanding this and other functions would be significantly easier (personally) if different logical blocks are preceded with brief descriptive comments. 
Though, in some cases a comment is overkill (such as when the logical block is small and self-explanatory) - but a new line would be ideal in helping with identifying it.

This practice, which I could not find a consistent name/documentation for, can be found everywhere in the binding controller for example.

Reply:
I have revised and extended the comments and factored the biggest function.